src/unix/lwt_domain.mli

```diff
@@ -0,0 +1,70 @@
+(* This file is part of Lwt, released under the MIT license. See LICENSE.md for
+   details, or visit https://github.com/ocsigen/lwt/blob/master/LICENSE.md. *)
+
+
+
+(** This module provides the necessary function ({!detach}) to schedule some
+    computations to be ran in parallel in a separate domain. The result of such
+    a computation is exposed to the caller of {!detach} as a promise. Thus, this
+    module allows to mix multicore parallelism with the concurrent-only
+    scheduling of the rest of Lwt. *)
+
+    val detach : ('a -> 'b) -> 'a -> 'b Lwt.t
+    (** [detach f x] runs the computation [f x] in a separate domain in
+        parallel.
+
+        [detach f x] evaluates to an Lwt promise which is pending until the
+        domain completes the execution of [f x] at which point it becomes
+        resolved. If [f x] raises an exception, then the promise is rejected.
+
+        If the task pool has not been initialised yet (see {!set_num_domains}),
+        then [detach] initializes it. The default number of domains is four (4).
+        four. It is recommended you initialise the task pool using
+        {!set_num_domains} with a number of domains equal to the number of
+        physical cores.
+
+        Note that the function [f] passed to [detach] cannot safely use {!Lwt}.
+        This is true even for implicit callback arguments (i.e.,
+        {!Lwt.with_value}). If you need to use {!Lwt} or interact with promises,
+        you must use {!run_in_main}. *)
+
+  val run_in_main : (unit -> 'a Lwt.t) -> 'a
+    (** [run_in_main f] can be called from a detached computation to execute [f
+        ()] in the parent domain, i.e. the one executing {!Lwt_main.run}.
+
+        [run_in_main f] blocks until [f ()] completes, then it returns its
+        result. If [f ()] raises an exception, [run_in_main f] raises the same
+        exception. The whole of {!Lwt} can be safely used from within [f].
+        However, note that implicit callback arguments are local to [f]. I.e.,
+        {!Lwt.get} can only retrieve values set inside of [f], and not those set
+        inside the promise that called [detach] that called [run_in_main].
+
+        Note that the calling domain will be idle until [f ()] completes
+        execution and returns the result. Thus, heavy use of [run_in_main] may
+        lead to most or all domains being frozen. It's also possible to create a
+        dead-lock when [run_in_main] is called (thus freezing a domain) with a
+        function that calls [detach] (thus needing a domain). Consequently, it
+        is recommended to use this function sparingly. *)
+
+  val set_num_domains : int -> unit
```

Comment:
Instead of `set_num_domains`, I would prefer to have `val setup_pool: int -> unit` and an explicit `val teardown_pool: unit -> unit`. 

I would also like `setup_pool` to raise an exception if a pool is currently active. I am not fond of `set_num_domains` implicitly shutting down the previous pool and creating a new pool. It may be the case that there are active tasks in the pool and implicitly tearing down pool and setting up a new one will silently kill those tasks. At least with an explicit `teardown_pool`, you get to choose to do the expensive operation explicitly. `setup_pool` and `teardown_pool` are what is used by `domainslib`. 

Also, an explict `teardown_pool` gives the chance for the program to cull the domains and not pay the associated cost of participating in the GC. This is useful if the program were to have a specific portion where there is parallelism opportunity and long portions where there isn't parallelism available.

Reply:
Done in c7e92da6ad4d9ab849f7f5b2dc7635d861efb6a5

Reply:
I'm wondering if, with this exception discipline, for consistency, we should not auto-initialise on `detach` and instead raise an exception there too. Would that make sense?

Reply:
Alternatively, when the pool isn't initialised, we could consider `max_domains = 1` and thus `detach` just runs within the main domain?

Reply:
> Alternatively, when the pool isn't initialised, we could consider max_domains = 1 and thus detach just runs within the main domain?

Here is how I thought about the problem. If the user does `Lwt_domain.detach` without setting up the pool, how likely is that they have forgotten to set up the pool? I would think very likely since they want to make their code go faster, but haven't informed the system about how many domains should be used. Hence, I would still prefer `detach` to raise an exception if a pool weren't set up. 

That said, I don't have a strong opinion about this, and it is ok to go for `detach` that auto initialises the pool.

Reply:
Alternatively, why not make `Lwt_detach` take an _explicit_ `pool` parameter to every computation, and have the init/teardown functions return a `pool` value?

This makes it possible for the application to assign computations to specific domain pools, which is coming up as a common requirement.

Reply:
I like that idea. The issue with that is that you'll have to thread through the `pool` value in your code. Porting your sequential Lwt code to use `Lwt_domain` will require modifying more code than the current API. Alternatively, if `domainslib` were to provide named pools, then the change can be local to where you're adding parallelism.

Reply:
Adding named pools into Domainslib does seem better, as it would also allow non-Lwt code to share the underlying domains (e.g. direct-style code)

Reply:
Passing an explicit pool parameter sounds like a good idea. This will also make managing pools easier when using multiple libraries like say Lwt_domain, domainslib and more that'd be developed in future with multicore features. In that case, can we use setup/teardown functions directly from domainslib? If offering the same API in `Lwt_domain` would make dependencies easier, we can have `Lwt_domain.setup_pool` return a domain pool and pass it around. It will need more code than the API in its current form, though this could be similar to domainslib APIs.

Reply:
I think the answer lies in which library 'owns' the scheduling and management of all the domains in an application. Typically it is a bad idea to multiplex schedulers since they will end up competing, so having a single place (like Domainslib) spawn and partition and schedule jobs to domains makes sense.  Once that core interface exists in Domainslib (e.g. via named pools as @kayceesrk suggests), then downstream libraries like Lwt and Async can pass the appropriate invocations to it.  

So we could have something like:

- `Domainslib`:
  * add support for naming pools with a number of sub-domains (the equivalent of `set_num_domains` and `setup_pool` and `teardown_pool` in this PR).
  * ensure there is a default pool with a sensible number of domains so if nothing is setup, there is still a pool to go to globally.
- `Lwt`:
  * add an interface `Lwt.detach ?pool fn` that can dispatch a computation to a pool optionally, but will go to the global Domainslib pool if none is specified.
  * this will be a simple interface by default (no pool required) but specialised applications that need per-pool dispatch (like the Tezos signature checkers, or MirageOS crypto libs) could setup a Domainslib named pool and pass in a `Lwt.detach ~pool` argument through their applications.

Later, other libraries like Async and Eio could also use the same Domainslib interface to create further named pools, but only the Domainslib scheduler would manage those pools, avoiding scheduler multiplexing.

Reply:
Note that for Lwt, it would make sense that the pool is created just before the call to `Lwt_main.run` (this is currently recommended in the documentation). This means that the threading of the `pool` value needs to start at the very top of the call graph of the whole program. Moreover, because `detach` is used for non-Lwt functions (its type is doesn't allow Lwt computations) the `pool` value need to be threaded all the way down to the deepest branches of the call graph of the whole program. Basically, it'd be very invasive.

So a default pool and/or some named pools would be a welcome addition.

-------------------------------------------------------

About calling `detach` without a prior `setup_pool`, I'm ok with the exception approach.

There is actually another reason why it might be good to go for the exception approach and it has to do with an under-documented feature(bug? who is to say?) of Lwt. Specifically, before the event-loop is started (when the call to `Lwt_main.run` is made) and after the event-loop has stopped (when the call to `Lwt_main.run` has returned), I/O jobs can be started and can even progress, but their promises can never be resolved.

With `Lwt_domain.detach` you can detach computations even if the event-loop is not running. I am not sure that the current mechanism allows for resolution of these promises when the scheduler is not running. (Currently it uses a feature of Lwt_unix that I am not very familiar with.)

In the future we might want to provide a solution not dependent on the Unix module, to provide solutions for other schedulers, etc. To this end, it makes more sense to have a simple interface with a simple semantic which we can have across all backends. And raising an exception if the pool is not set up is a simple interface.

Reply:
https://github.com/ocaml-multicore/domainslib/pull/45 adds named pools to domainslib. This enables obtaining pools at a later point after their creation with a "name" attached to it during its creation. Essentially this eliminates the need to thread through pool. 

With this, we can eliminate the global pool present in `Lwt_domain` and pass pool as an argument to `detach`. We can also provide wrappers for initializing/tearing down pools in the interface itself, so that it's self contained. I plan to switch to this soon, if there are concerns about this please do let us know.

Reply:
the PR is rebased to [domainslib.0.3.1](https://github.com/ocaml-multicore/domainslib/releases/tag/0.3.1) (cf. 02ab32f933cf798bc5ffe56b1d7587d90ec1c10c). Major updates are:
* A `pool` is passed explicitly to `detach` computations. This eliminates the need for a global pool we had earlier. If the pool is not active, it raises an `Invalid_argument` exception.

* domainslib `0.3.1` has support for named pools which enables obtaining pools with a name at a later point, and hence we don't need to thread through the pool argument.

* Includes wrappers to some domainslib functions like `setup_pool`, `teardown_pool` etc. to make this module self-contained.

Reply:
All the points discussed in this conversation have been addressed. I'm resolving the convesation.